[unitaryhack] Add SPSA optimization

ixime opened the pull request:
**Context:**
Adding a feature as part of unitaryhack related to issue #2451

**Description of the Change:**
The qml.SPSAOptimizer was added with its corresponding tests based on: 

J. Spall, "An Overview of the Simultaneous Perturbation Method for Efficient Optimization." Johns Hopkins api technical digest, 19(4), 482-492, 1998

**Benefits:**
* The Simultaneous Perturbation Stochastic Approximation method (SPSA) is an iterative algorithm for optimization ideal for noisy loss functions.
* In contrast to other methods that perform multiple operations to determine the gradient, SPSA only measures two times the loss function to obtain it.

**Possible Drawbacks:**

**Related GitHub Issues:**
Closes https://github.com/PennyLaneAI/pennylane/issues/2451

ixime commented:
I think I finished everything, but I am not sure about the cases of hybrid classical-quantum workflows. as I commented in #2451

antalszava commented:
Hi @ixime, thank you, we'll be reviewing the PR soon. :slightly_smiling_face:

A small note: there's another PR (https://github.com/PennyLaneAI/pennylane/pull/2659) targeting the [same issue](https://github.com/PennyLaneAI/pennylane/issues/2451). There is the opportunity of teaming up and creating a single solution together for the same issue. This is optional, just wanted to let you know. :slightly_smiling_face:

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+    ...                       [1.25907603, 0.40854088]])
+    >>> tensor_out = tf.Variable([0,0])
+
+    >>> init = tf.compat.v1.global_variables_initializer()
+    >>> with tf.compat.v1.Session() as sess:
```
Note two things here:
* Usually PennyLane optimizers (such as `qml.SPSAOptimizer`) can not be used with other interfaces ([see here](https://pennylane--2661.org.readthedocs.build/en/2661/introduction/interfaces/tf.html#optimization-using-tensorflow)) - here it might work well because we SPSA uses the results of cost function evaluations;
* We would like to refrain from using TensorFlow v1 as PennyLane only support TF v2.

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+
+
+    Keyword Args:
+        maxiter (int): the maximum number of iterations expected to be performed
```
Worth noting that this arg is used to determine `A` under the hood and if `A` is supplied then this arg is ignored.

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+
+    Keyword Args:
+        maxiter (int): the maximum number of iterations expected to be performed
+        alpha (float): An hyperparameter to calculate :math:`a_k=\frac{a}{(A+k+1)^\alpha}`
```
```suggestion
        alpha (float): A hyperparameter to calculate :math:`a_k=\frac{a}{(A+k+1)^\alpha}`
```

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+    Keyword Args:
+        maxiter (int): the maximum number of iterations expected to be performed
+        alpha (float): An hyperparameter to calculate :math:`a_k=\frac{a}{(A+k+1)^\alpha}`
+            for each iteration. Its asymptotically optimal value is 1.0
```
```suggestion
            for each iteration. Its asymptotically optimal value is 1.0.
```

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+        alpha (float): An hyperparameter to calculate :math:`a_k=\frac{a}{(A+k+1)^\alpha}`
+            for each iteration. Its asymptotically optimal value is 1.0
+        gamma (float): An hyperparameter to calculate :math:`c_k=\frac{c}{(k+1)^\gamma}`
+            for each iteration. Its asymptotically optimal value is 1/6
```
```suggestion
            for each iteration. Its asymptotically optimal value is 1/6.
```

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+            for each iteration. Its asymptotically optimal value is 1.0
+        gamma (float): An hyperparameter to calculate :math:`c_k=\frac{c}{(k+1)^\gamma}`
+            for each iteration. Its asymptotically optimal value is 1/6
+        c (float): An hyperparameter related to the expected noise. It should be
```
```suggestion
        c (float): A hyperparameter related to the expected noise. It should be
```

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+            for each iteration. Its asymptotically optimal value is 1/6
+        c (float): An hyperparameter related to the expected noise. It should be
+            approximately the standard deviation of the expected noise on the cost function
+        A (float): The stability constant expected to be 10% of maximum number
+            of expected iterations
```
```suggestion
        A (float): The stability constant expected to be 10% of the maximum number
            of expected iterations.
```

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+            approximately the standard deviation of the expected noise on the cost function
+        A (float): The stability constant expected to be 10% of maximum number
+            of expected iterations
+        a (float): An hyperparameter expected to be small in noisy situations,
```
```suggestion
        a (float): A hyperparameter expected to be small in noisy situations,
```

antalszava replied:
Worth rephrasing `noisy situations`.

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+        self.ak = self.a / (self.A + 1 + 1.0) ** self.alpha
+
+    def step_and_cost(self, objective_fn, *args, **kwargs):
+        """Update the parameter array :math:`x` with one step of the optimizer and return
```
Worth updating the $x$ notation in this docstring.

MoritzWillmann replied:
Do you mean changing it to theta?

antalszava replied:
Yes :)

antalszava commented on doc/releases/changelog-dev.md:
````diff
@@ -180,6 +180,32 @@
   tensor([0.69301172, 0.67552491, 0.65128847], requires_grad=True)
   ```
 
+* Added new optimizer `qml.SPSAOptimizer` for noisy problems. It performs less device executions, due
+to the fact that it performs two measures to determine the gradient.
````
Should be okay to be brief here and link to the literature. More info can be shared in 2-3 sentences below the link, before the example.
```suggestion
* Added the new optimizer, `qml.SPSAOptimizer` that implements the simultaneous perturbation stochastic approximation method based on
  [An Overview of the Simultaneous Perturbation Method for Efficient Optimization](https://www.jhuapl.edu/SPSA/PDF-SPSA/Spall_An_Overview.PDF).
```

antalszava commented on doc/releases/changelog-dev.md:
```diff
@@ -180,6 +180,32 @@
 
+* Added new optimizer `qml.SPSAOptimizer` for noisy problems. It performs less device executions, due
+to the fact that it performs two measures to determine the gradient.
+[(#2661)](https://github.com/PennyLaneAI/pennylane/pull/2661)
```
```suggestion
  [(#2661)](https://github.com/PennyLaneAI/pennylane/pull/2661)
```

antalszava commented on doc/releases/changelog-dev.md:
````diff
@@ -180,6 +180,32 @@
+to the fact that it performs two measures to determine the gradient.
+[(#2661)](https://github.com/PennyLaneAI/pennylane/pull/2661)
+
+```python
+dev = qml.device("default.qubit", wires=1)
````
Note that this code block needs to be indented with 2 space characters (see other examples in this file).

antalszava commented on doc/releases/changelog-dev.md:
```diff
@@ -180,6 +180,32 @@
+
+def exp_val_circuit(params):
+    circuit(params, range(dev.num_wires))
+    return qml.expval(h2_ham)
```
`h2_ham` is not defined - would be good to have a def for it such that the example is runnable simply (or the observable may be different).

antalszava commented on doc/releases/changelog-dev.md:
```diff
@@ -180,6 +180,32 @@
+    circuit(params, range(dev.num_wires))
+    return qml.expval(h2_ham)
+
+params = np.random.normal(0, np.pi, (num_qubits, 3), requires_grad=True)
+cost = qml.QNode(exp_val_circuit, dev)
+
+max_iterations = 100
+opt = qml.SPSAOptimizer(maxiter=max_iterations)
+for n in range(max_iterations):
+    params, energy = opt.step_and_cost(cost, params)
```
Could be nice to show that the parameters are updated by adding a `pycon` block too to this example.

antalszava commented on tests/optimize/test_spsa.py:
```diff
@@ -0,0 +1,332 @@
+    @pytest.mark.parametrize("args", [0, -3, 42])
+    @pytest.mark.parametrize("f", univariate)
+    def test_apply_grad(self, args, f):
+        """
+        Test that a gradient step can be applied correctly with a univariate
+        function.
+        """
```
This formatting suggestion applies to other test case docstrings too.
```suggestion
        """Test that a gradient step can be applied correctly with a univariate
        function."""
```

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+                delta[index] = np.random.choice([-1, 1], size=arg.shape)
+                thetaplus[index] = arg + ck * delta[index]
+                thetaminus[index] = arg - ck * delta[index]
+        yplus = objective_fn(*thetaplus, **kwargs)
+        yminus = objective_fn(*thetaminus, **kwargs)
```
What happens if `objective_fn` returns a vector?

MoritzWillmann replied:
Is this defined?

antalszava replied:
The linked article considers scalar functions. Nonetheless, there may be extensions to vector-valued functions too.

My question here is more so practical: what happens with the machinery of `qml.SPSAOptimizer` if `objective_fn` *does* return a vector?

It may be a common use case that a user defines a QNode that returns `qml.probs(wires=wires)` for some `wires` and would like to use `qml.SPSAOptimizer`. Does the code break? Is an error being raised? If so, does it provide a descriptive error message? This should be a case tested. :+1:

ixime replied:
So for now only scalar functions are enough?

antalszava replied:
Yes, that's okay. See that there's a validation check now too. :+1:

antalszava commented on tests/optimize/test_spsa.py:
```diff
@@ -0,0 +1,332 @@
+            qml.RX(variables[0], wires=[0])
+            qml.RY(variables[1], wires=[0])
+            qml.RY(variables[2], wires=[0])
+            return qml.expval(qml.PauliZ(0))
```
Definitely worth adding an example for using `qml.Hamiltonian`.

antalszava reviewed (changes requested):
Hi @ixime, thank you for the contribution! :pray: 

The overall base of the logic is looking great! :slightly_smiling_face: There's room for improvements here and there, left some comments. I'll be providing some more comments tomorrow, thought of submitting the ones I've made already in the meantime as some might take time to address.

MoritzWillmann commented:
@ixime my discord is moritz#8621. Tell me what you need me to do here :smile:

MoritzWillmann commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+                :math:`\hat{g}_k(\hat{\theta}_k)` and ``None``
+        """
+        self.increment_k()
+        ck = self.c / (self.k + 1.0) ** self.gamma
```
I think the version used by @ixime is from the "Implementation of the Simultaneous Perturbation Algorithm for
Stochastic Optimization" paper but I agree initializing to 1 would improve understanding

antalszava replied:
Got it :) Either should work well, just noticed the addition by `1.0` which can feel a bit out of place when reading the code, but nothing major.

MoritzWillmann commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+        for index, arg in enumerate(args):
+            if getattr(arg, "requires_grad", False):
+                delta[index] = np.random.choice([-1, 1], size=arg.shape)
+                thetaplus[index] = arg + ck * delta[index]
+                thetaminus[index] = arg - ck * delta[index]
```
Do you think the computational savings outweigh the loss in readability?

antalszava replied:
Could do
```py
multiplier = ck * delta[index]
thetaplus[index] = arg + multiplier
thetaminus[index] = arg - multiplier
```
which I guess should be okay for readability too. Or do something comparable.

MoritzWillmann commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,255 @@
+        """
+        self.increment_k()
+        ck = self.c / (self.k + 1.0) ** self.gamma
+        delta = list(args)
```
Yes, it's needed later on to calculate the approximated gradient. I'd argue though, that it should be initialized to an empty list, rather than copying args and omit the initialization of grad to an empty list.

MoritzWillmann commented:
Hey @ixime, i resolved some comments. Can you give me rights to push to your clone so I can commit the updates? Thanks!

ixime commented:
> 

done! :smile:

MoritzWillmann commented:
How are we feeling about adding the ability to pass the hyperparameters `a`, `A`, `alpha`, `c` and `gamma` to the `step` and `step_and_cost` functions? This could simplify the process of adapting them on the go.

ankit27kh commented:
Also, will this be working with `lightning.qubit`, `lightning.gpu`, `default.mixed`?

MoritzWillmann commented:
Hey @ankit27kh, thank you so much for your input!

> how are you verifying your results?

So far, only if the returned result for the objective function is correct. Checking a random estimated gradient is hard. Do you have any suggestions how to check it? Maybe checking for convergence in the case of a simple quadratic program could be an Idea. What do you think, @ixime?

> Can you also check the results against Qiskit and noisyopt's SPSA implementation? That we are getting similar results with similar values.

I don't know If that's desirable for a unit test. (cc @antalszava)

> And also, check for the number of device executions.

Great Idea, though that should be clear from implementation.

> Also, will this be working with lightning.qubit, lightning.gpu, default.mixed?

I don't see no reason why it shouldn't. We can definitely add test cases concerning this.

ankit27kh commented:
> I don't know If that's desirable for a unit test.

Hey @MoritzWillmann, I don't mean as a unit test. Just as a test for confirmation, you can do it once and put the results here.
This will verify results and implementation while also saying something about the interoperability of different libraries.

antalszava commented:
> > Can you also check the results against Qiskit and noisyopt's SPSA implementation? That we are getting similar results with similar values.

> I don't know If that's desirable for a unit test. (cc @antalszava)

I would agree. We wouldn't like to depend on the logic of another library to check correctness in our test suite. Having said that, it can serve as a sanity check separate from this PR, but making it a part of our test suite would introduce a new dependency and give a (potentially) false sense of security. Even in our test suite, we at times come across cases where simulation results that are used for checking correctness turn out to be incorrect.

> I don't see no reason why it shouldn't. We can definitely add test cases concerning this.

That would be ideal, indeed. :slightly_smiling_face: These devices should be available in the test suite to be run.

---

(Thank you for addressing the comments already! I'll be checking in shortly for another pass of the PR to provide thoughts in the form of comments. Great to see the collaboration here! :blush:)

antalszava commented on tests/optimize/test_spsa.py:
```diff
@@ -0,0 +1,337 @@
+
+
+class TestSPSAOptimizer:
+    """Test the SPSA optimizer"""
```
It would be good to first test that the various attributes are set correctly on the `optimizer` object.

antalszava commented on tests/optimize/test_spsa.py:
```diff
@@ -0,0 +1,337 @@
+        spsa_opt = qml.SPSAOptimizer(maxiter=10)
+        args = np.array([args], requires_grad=True)
+
+        alpha = 0.602
+        gamma = 0.101
+        c = 0.2
```
Let's use non-default values too (e.g., by parametrizing these tests to have another value too).

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,282 @@
+        A (float): The stability constant expected to be 10% of the maximum number
+            of expected iterations.
+        a (float): A hyperparameter expected to be small in noisy situations,
+            whose value could be :math:`\frac{mag(\Delta\theta)}{mag(g(\theta))}(A+1)^\alpha`
```
Where does `\frac{mag(\Delta\theta)}{mag(g(\theta))}(A+1)^\alpha` come from? If it's not mentioned in the Overview in the docstring, then it's worth citing another source.

ixime replied:
I've added the reference in which Spall makes some recommendations about constants and gains

antalszava commented on tests/optimize/test_spsa.py:
```diff
@@ -0,0 +1,337 @@
+        k = 1
+        ck = c / (k + 1.0) ** gamma
+        ak = a / (A + k + 1.0) ** alpha
+        tol = np.maximum(np.abs(f(args - ck)), np.abs(f(args + ck)))
```
Why is the tolerance chosen as such?

ixime replied:
Yes, it was wrong. I've already changed those examples. However, please tell me what you think of the new version.

antalszava commented on tests/optimize/test_spsa.py:
```diff
@@ -0,0 +1,337 @@
+
+        assert np.all(res == expected)
+
+    def test_step_and_cost_supplied_cost3(self):
```
```suggestion
    def test_step_and_cost_hamiltonian(self):
```

antalszava commented on tests/optimize/test_spsa.py:
```diff
@@ -0,0 +1,337 @@
+        expected = f(args)
+        assert np.all(res == expected)
+
+    def test_step_and_cost_supplied_cost2(self):
```
Worth updating the test names such that it describes the test case more precisely. That way no numbers would have to be used.

antalszava reviewed (commented):
Posting a couple of comments. :tada: Overall this is looking awesome, could just do with a bit more polishing.

For further device tests, most mentioned devices (`default.qubit`, `lightning.qubit` and `default.mixed`) could indeed be tested. Testing `lightning.gpu` in this repo might be more challenging, so it's okay to leave that out.

ixime commented:
> 

I put them on compute_grad() in case for some reason you just want to compute the grad

ixime commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,282 @@
+        Therefore, in case of using ``step_and_cost`` method instead of ``step``, the number
+          of executions will include the cost function evaluations.
+
+    .. note::
+
+        In cases of hybrid classical-quantum workflows:
+
+        * In implementation of a QNode as a layer of a Keras sequential model,
+          possible optimizers for the model are from the classical platform i.e.
+          `tf.keras.optimizers.SGD
+          <https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/SGD>`_.
+        * In a hybrid classical-quantum-classical workflow where we use SPSAOptimizer
+          for the quantum part, we have to extract the values of the classical tensor
+          in order to use it in the quantum circuit as inputs and assign the output
+          of the quantum circuit to a subsecuent classical tensor.
+
+
```
These are the comments I've added in relation to hybrid classical-quantum workflows. At first I wasn't sure exactly what was needed in relation to that point. So maybe they can go.

ixime commented:
> How are we feeling about adding the ability to pass the hyperparameters `a`, `A`, `alpha`, `c` and `gamma` to the `step` and `step_and_cost` functions? This could simplify the process of adapting them on the go.

I put them on compute_grad() in case for some reason you just want to compute the grad

ixime commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,282 @@
+        self.ak = self.a / (self.A + 1 + 1.0) ** self.alpha
+
+    def step_and_cost(self, objective_fn, *args, **kwargs):
+        r"""Update the parameter array :math:`\hat{\theta}_k` with one step of the optimizer and return
```
that's odd. I did make docs and it rendered ok
![Screenshot at 2022-06-07 23-17-33](https://user-images.githubusercontent.com/25181125/172530796-1a9cd62e-3ab7-489f-ad54-a87f7731a08f.png)
.

ixime commented:
> > > Can you also check the results against Qiskit and noisyopt's SPSA implementation? That we are getting similar results with similar values.
> 
> > I don't know If that's desirable for a unit test. (cc @antalszava)
> 
> I would agree. We wouldn't like to depend on the logic of another library to check correctness in our test suite. Having said that, it can serve as a sanity check separate from this PR, but making it a part of our test suite would introduce a new dependency and give a (potentially) false sense of security. Even in our test suite, we at times come across cases where simulation results that are used for checking correctness turn out to be incorrect.
>
So far I've compared `qml.SPSAOptimizer` with noisyopt's implementation and its very similar. However their implementation does two executions less than ours. 
> > I don't see no reason why it shouldn't. We can definitely add test cases concerning this.
> 
> That would be ideal, indeed. slightly_smiling_face These devices should be available in the test suite to be run.
> 
> (Thank you for addressing the comments already! I'll be checking in shortly for another pass of the PR to provide thoughts in the form of comments. Great to see the collaboration here! blush)

MoritzWillmann commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,282 @@
+    * :math:`k` is the current iteration step,
+    * :math:`\hat{\theta}_k` are the input parameters at iteration step :math:`k`,
+    * :math:`y` is the objective function,
+    * :math:`c_k=\frac{c}{(k+1)^\gamma}` is the gain sequence corresponding to evaluation step size and
```
yes, missed this one. Thanks!

ixime commented:
> light
I forgot to mention that I added some tests for ``lightning.qubit`` and ``default.mixed``

antalszava commented:
@ixime thank you! :) Just to confirm, should I go for a re-review on the entire PR?

ixime commented:
> @ixime thank you! :) Just to confirm, should I go for a re-review on the entire PR?

yes, please!

MoritzWillmann commented:
Hey @ixime, sorry I couldn't help yesterday. Do you need me to do anything else?

ixime commented:
> Hey @ixime, sorry I couldn't help yesterday. Do you need me to do anything else?

No problem. :smile: I think I covered all the comments, or do you see something misses? Did you compare the optimizer to qiskit's ? I only compared it to noisyopt's implementation

ankit27kh commented:
Hey @ixime, quick question, the `max_iter` and `A` hyperparameters seem to be connected as the latter is fixed depending on the former in cases where the latter is not supplied.  `max_iter` has the default value of 200. How much difference does it make to not have defined `max_iter` correctly during initialization? For example:
```
opt = qml.SPSAOptimizer()
for _ in range(epochs):
    params = opt.step(cost, params)
```
What effect will this have in my optimization process if I'm using 10/20 or 1000 epochs but `max_iter` gets set to 200?
Is the effect big enough to make `max_iter` a required argument?

MoritzWillmann commented:
> Did you compare the optimizer to qiskit's ? I only compared it to noisyopt's implementation
Not yet, but I can do it :smile:

ixime commented:
> 

On the other hand, it will always have smaller steps on 1000+ iterations, unless we set other value for `A`

ankit27kh commented:
> > Did you compare the optimizer to qiskit's ? I only compared it to noisyopt's implementation
> 
> Not yet, but I can do it smile

Hey @MoritzWillmann, please do this comparison too. I'm using both noisyopt and Qiskit currently and do not want any surprises when I switch too this implementation! 😆 😅

ankit27kh commented:
> > 
> 
> On the other hand, it will always have smaller steps on 1000+ iterations, unless we set other value for `A`

Yes so in my example if I do not set `max_iter` or `A`, `A` will be set to `200*0.1`
Then if I use more or fewer epochs, the result will be different than the case where I explicitly set `max_iter` to the number of epochs.
If this difference is big, maybe setting this value should be required.

MoritzWillmann commented:
Maybe we should then also check if k is larger than max_iter and warn if that's the case

ixime commented:
> Maybe we should then also check if k is larger than max_iter and warn if that's the case

I don't know, because again, maybe you do want an exponential decrease of `ak`

MoritzWillmann commented:
> I don't know, because again, maybe you do want an exponential decrease of `ak`

But wouldn't this kind of control be better suited with setting `A` ? I feel like It could be unintuitive if you can/must control something by setting a false `max_iter`.

ankit27kh commented:
> > I don't know, because again, maybe you do want an exponential decrease of `ak`
> 
> But wouldn't this kind of control be better suited with setting `A` ? I feel like It could be unintuitive if you can/must control something by setting a false `max_iter`.

`max_iter` is much more user friendly than `A`. An advance user will probably define all hyperparameters anyway. For simple use, I think `max_iter` is more intuitive to understand.

MoritzWillmann commented:
> > > Did you compare the optimizer to qiskit's ? I only compared it to noisyopt's implementation
> > 
> > 
> > Not yet, but I can do it smile
> 
> Hey @MoritzWillmann, please do this comparison too. I'm using both noisyopt and Qiskit currently and do not want any surprises when I switch too this implementation! 😆 😅

I yield similar results for a couple of different functions. I think you're good to go 😆

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,285 @@
+        if yplus.size > 1:
+            raise ValueError(
+                "The objective function must be a scalar function for the gradient "
+                "to be computed. "
```
```suggestion
                "to be computed."
```

antalszava commented on tests/optimize/test_spsa.py:
```diff
@@ -0,0 +1,466 @@
+"""Tests for the SPSA optimizer"""
+import os
+
+os.environ["OMP_NUM_THREADS"] = "2"
```
Why was this added?

ixime replied:
I was testing the parallelization options, but is not necessarily.

antalszava commented on tests/optimize/test_spsa.py:
```diff
@@ -0,0 +1,466 @@
+    def test_apply_grad(self, args, f):
+        """Test that a gradient step can be applied correctly with a univariate
+        function."""
+        spsa_opt = qml.SPSAOptimizer(maxiter=10, c=0.1, gamma=0.3)
+        args = np.array([args], requires_grad=True)
+
+        alpha = 0.602
+        gamma = 0.3
+        c = 0.1
```
Minor: to improve readability
```suggestion
        gamma = 0.3
        c = 0.1
        spsa_opt = qml.SPSAOptimizer(maxiter=10, c=c, gamma=gamma)
        args = np.array([args], requires_grad=True)

        alpha = 0.602
```

antalszava commented on tests/optimize/test_spsa.py:
```diff
@@ -0,0 +1,466 @@
+        ak = a / (A + k) ** alpha
+        deltas = np.array(np.meshgrid([1, -1], [1, -1])).T.reshape(-1, 2)
+
+        spsa_opt = qml.SPSAOptimizer(maxiter=10, A=10)
```
Minor for readability
```suggestion
        spsa_opt = qml.SPSAOptimizer(maxiter=10, A=A)
```

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -170,46 +171,47 @@ def __init__(self, maxiter=200, alpha=0.602, gamma=0.101, c=0.2, A=None, a=None)
-        the step and the corresponding objective function.
+        r"""Update the parameter array :math:`\hat{\theta}_k` with one step of the
+        optimizer and return the step and the corresponding objective function. The number
+        of steps is being counted through calls to ``step_and_cost`` and ``cost``.
```
```suggestion
        of steps stored by the ``k`` attribute of the optimizer is counted internally when calling ``step_and_cost`` and ``cost``.
```

antalszava commented on pennylane/optimize/spsa.py:
```diff
@@ -0,0 +1,292 @@
+
+    .. note::
+
+        * The number of quantum device executions is :math:`2*iter*num\_terms\_hamiltonian`.
```
```suggestion
        * One SPSA iteration step of a cost function that involves computing the expectation value of
        a Hamiltonian with ``M`` terms requires :math:`2*M` quantum device executions.
```

antalszava commented on tests/optimize/test_spsa.py:
```diff
@@ -0,0 +1,501 @@
+        assert res == params
+
+    def test_obj_func_with_probs_not_a_scalar_function(self):
+        """Test that if the objective function is not a
+        scalar function like qml.probs(), an error is raised."""
```
```suggestion
        """Test that if the objective function is not a
        scalar function, like qml.probs(), an error is raised."""
```

antalszava reviewed (changes requested):
Hi @ixime and @MoritzWillmann! This is looking amazing overall. :clap: :fireworks:

The suggestions I have this time are smaller. Once they are addressed, I'll make another pass, but the overall addition here is very near approval.

ixime commented:
> 

I tried to solve the conflicts, but I don't have write access to solve them. My fork master is up to date and apparently I merge successfully my local branch with my up to date local master :woman_shrugging:  I am little lost here.. did I miss some step?

josh146 commented:
@ixime It should be fixed now :)

antalszava commented on tests/optimize/test_spsa.py:
```diff
@@ -0,0 +1,510 @@
+        ):
+            spsa_opt.step(f, args)
+
+    def test_obj_func_not_a_scalar_function_with_tensorflow_interface(self):
```
```suggestion
    @pytest.mark.tf
    def test_obj_func_not_a_scalar_function_with_tensorflow_interface(self):
```

ixime commented:
Hi @antalszava why some times the codecov report is fine and the next check is all wrong without changing a bit on the code, just the docstring?

antalszava reviewed (approved):
Looks good @ixime, @MoritzWillmann! :tada: :fireworks: This is an amazing and high-quality addition, congrats :slightly_smiling_face: :clap: 

Adding the Unitary HACK accepted label too.

> Hi @antalszava why some times the codecov report is fine and the next check is all wrong without changing a bit on the code, just the docstring?

I think the issue came from a previous state where one test was failing and it didn't report to Codecov correctly. Now it should be okay :+1:
